components/sidebar/sidebar.jsx

```diff
@@ -0,0 +1,761 @@
+
+        actions: PropTypes.shape({
+            goToChannelById: PropTypes.func.isRequired
+        }).isRequired
+    }
+
+    static defaultProps = {
+        currentChannel: {}
+    }
+
+    constructor(props) {
+        super(props);
+
+        this.badgesActive = false;
+        this.firstUnreadChannel = null;
+        this.lastUnreadChannel = null;
+
+        this.isLeaving = new Map();
+        this.isSwitchingChannel = false;
+        this.closedDirectChannel = false;
+
+        this.state = {
+            newChannelModalType: '',
+            showDirectChannelsModal: false,
+            showMoreChannelsModal: false
+        };
+    }
+
+    componentDidMount() {
+        this.updateUnreadIndicators();
+
+        document.addEventListener('keydown', this.navigateChannelShortcut);
+        document.addEventListener('keydown', this.navigateUnreadChannelShortcut);
+    }
+
+    componentWillUpdate() {
+        this.updateUnreadIndicators();
+    }
+
+    componentDidUpdate(prevProps) {
+        // if the active channel disappeared (which can happen when dm channels autoclose), go to town square
+        if (this.props.currentTeam === prevProps.currentTeam &&
+            this.props.currentChannel.id === prevProps.currentChannel.id &&
+            !this.channelIdIsDisplayedForProps(this.props, this.props.currentChannel.id) &&
+            this.channelIdIsDisplayedForProps(prevProps, this.props.currentChannel.id)
+        ) {
+            this.closedDirectChannel = true;
+            browserHistory.push('/' + this.props.currentTeam.name + '/channels/town-square');
+            return;
+        }
+
+        if (!Utils.isMobile()) {
+            $('.sidebar--left .nav-pills__container').perfectScrollbar();
+        }
+
+        // reset the scrollbar upon switching teams
+        if (this.props.currentTeam !== prevProps.currentTeam) {
+            this.refs.container.scrollTop = 0;
+            $('.nav-pills__container').perfectScrollbar('update');
+        }
+
+        // close the LHS on mobile when you change channels
+        if (this.props.currentChannel.id !== prevProps.currentChannel.id) {
+            if (this.closedDirectChannel) {
+                this.closedDirectChannel = false;
+            } else {
+                $('.app__body .inner-wrap').removeClass('move--right');
+                $('.app__body .sidebar--left').removeClass('move--right');
+                $('.multi-teams .team-sidebar').removeClass('move--right');
+            }
+        }
+
+        this.updateTitle();
+
+        this.setBadgesActiveAndFavicon();
+        this.setFirstAndLastUnreadChannels();
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener('keydown', this.navigateChannelShortcut);
+        document.removeEventListener('keydown', this.navigateUnreadChannelShortcut);
+    }
+
+    setBadgesActiveAndFavicon() {
+        this.lastBadgesActive = this.badgesActive;
+
+        this.getDisplayedChannels().map((channelId) => {
```

Comment:
Finally I have realized I don't need the map at all, because the check of existence of at least a membership comes from the old code, and I think is not needed to show badge in the title.